[ISSUE] Issue with `databricks_permissions` resource since `1.10.0`

### Configuration
We have many `databricks_permissions` resources in our templates. They all look similar to the sample code below
```hcl
resource "databricks_notebook" "userlogs" {
  content_base64 = base64encode(<<-EOT
    logs = spark.read.option("header", True).csv("/mnt/UserLogs/userlogs_notebook.csv")
    display(logs)
    EOT
  )
  path     = "/Workspace/Audit_Logs/UserLogs_Notebook"
  language = "PYTHON"
}

resource "databricks_permissions" "userlogs" {
  directory_path = "/Workspace/Audit_Logs"

  access_control {
    group_name       = "workspace_admins"
    permission_level = "CAN_EDIT"
  }

  depends_on = [
    databricks_notebook.userlogs,
  ]
}
```

### Expected Behavior
Before `1.10.0`, the templates can be applied without any errors. All the resources were created successfully.

### Actual Behavior
Since `1.10.0`, Terraform failed to apply and showed the following errors (please ignore the prefix from the logs as they were obtained from my CI/CD system)
```
14:11:48   Info     |       module.workspace_team_data_science.databricks_permissions.cluster_usage[0]: Refreshing state... [id=/clusters/0708-013148-c1p0ylun]
14:11:48   Info     |       module.workspace_project_breach_reporting_data_publication.databricks_permissions.cluster_usage[0]: Refreshing state... [id=/clusters/0706-073051-s5qacrut]
14:11:48   Info     |       module.workspace_project_breach_reporting_data_publication.aws_ssm_parameter.cluster_id: Refreshing state... [id=/preprod/datalake/databricks/cluster-id/project-stratpolbrreportingdatapubl]
14:11:48   Info     |       module.workspace_project_audit_logs.aws_ssm_parameter.cluster_id: Refreshing state... [id=/preprod/datalake/databricks/cluster-id/project-auditlogs]
14:11:48   Info     |       module.s3_mount.databricks_job.s3_mount: Refreshing state... [id=270265880191491]
14:12:03   Error    |       Error: Get "https://asic-preprod.cloud.databricks.com/api/2.0/preview/scim/v2/Me": EOF
14:12:03   Error    |       with module.workspace_team_strategic_intelligence.databricks_permissions.cluster_usage[0],
14:12:03   Error    |       on modules/workspace/databricks_clusters.tf line 61, in resource "databricks_permissions" "cluster_usage":
14:12:03   Error    |       61: resource "databricks_permissions" "cluster_usage" {
14:12:03   Error    |       Error: Get "https://asic-preprod.cloud.databricks.com/api/2.0/preview/scim/v2/Me": EOF
14:12:03   Error    |       with module.workspace_team_wealth_management.databricks_permissions.cluster_usage[0],
14:12:03   Error    |       on modules/workspace/databricks_clusters.tf line 61, in resource "databricks_permissions" "cluster_usage":
14:12:03   Error    |       61: resource "databricks_permissions" "cluster_usage" {
14:12:03   Error    |       Error: Get "https://asic-preprod.cloud.databricks.com/api/2.0/preview/scim/v2/Me": EOF
14:12:03   Error    |       with module.workspace_team_nlp_registered_liquidators.databricks_permissions.cluster_usage[0],
14:12:03   Error    |       on modules/workspace/databricks_clusters.tf line 61, in resource "databricks_permissions" "cluster_usage":
14:12:03   Error    |       61: resource "databricks_permissions" "cluster_usage" {
14:12:03   Error    |       Error: Get "https://asic-preprod.cloud.databricks.com/api/2.0/preview/scim/v2/Me": EOF
14:12:03   Error    |       with module.workspace_team_data_analytics_platforms.databricks_permissions.workspace[0],
14:12:03   Error    |       on modules/workspace/databricks_notebooks.tf line 13, in resource "databricks_permissions" "workspace":
14:12:03   Error    |       13: resource "databricks_permissions" "workspace" {
14:12:03   Error    |       Error: cannot read permissions: Get "https://asic-preprod.cloud.databricks.com/api/2.0/preview/scim/v2/Me": EOF
14:12:03   Error    |       with module.workspace_ml_reportable_situations.module.workspace_publish.databricks_permissions.workspace[0],
14:12:03   Error    |       on modules/workspace/databricks_notebooks.tf line 13, in resource "databricks_permissions" "workspace":
14:12:03   Error    |       13: resource "databricks_permissions" "workspace" {
14:12:03   Error    |       Error: cannot read permissions: Get "https://asic-preprod.cloud.databricks.com/api/2.0/preview/scim/v2/Me": EOF
14:12:03   Error    |       with module.workspace_project_information_resource_centre.databricks_permissions.workspace[0],
14:12:03   Error    |       on modules/workspace/databricks_notebooks.tf line 13, in resource "databricks_permissions" "workspace":
14:12:03   Error    |       13: resource "databricks_permissions" "workspace" {
14:12:03   Error    |       Error: Get "https://asic-preprod.cloud.databricks.com/api/2.0/preview/scim/v2/Me": EOF
14:12:03   Error    |       with module.workspace_project_asic_property.databricks_permissions.userlogs[0],
14:12:03   Error    |       on modules/workspace/databricks_notebooks.tf line 38, in resource "databricks_permissions" "userlogs":
14:12:03   Error    |       38: resource "databricks_permissions" "userlogs" {
```

### Steps to Reproduce
1. `terraform init`
2. `terraform plan` or ` terraform apply`.

### Terraform and provider versions
- Terraform: `1.3.7`
- Databricks: `1.10.0`

### Debug Output
Nothing standout from the debug logs. All the error messages were similar to the output above.

### Important Factoids
Looks like the migration to the Databricks SDK has changed the way the SCIM API was used.

I am facing a similar issue with referencing a when provisioning access to cluster polices.

Issue occurs only on 1.10.0 and 1.10.1, I have upgraded as I am seeing other issues that I was trying to reproduce on the latest version before raising an issue. 

I am available 10am-5pm (UTC+10) if you would like chat in real time. 
example code 
```
resource "databricks_cluster_policy" "cluster_policy" {
  name       = var.CLUSTER_POLICY_NAME
  definition = jsonencode(var.CLUSTER_POLICY_DEF)
}

resource "databricks_permissions" "can_use_cluster_policyinstance_profile" {
  cluster_policy_id = databricks_cluster_policy.cluster_policy.id
  access_control {
    group_name       = var.POLICY_USER_GROUP
    permission_level = "CAN_USE"
  }
  depends_on = [
    databricks_cluster_policy.cluster_policy
  ]
}
```


I am using the following providers

```
 azurerm = {
      source  = "hashicorp/azurerm"
      version = "3.45.0"
    }
    databricks = {
      source  = "databricks/databricks"
      version = "1.10.1"

    }
```

```
2023-03-01T04:53:20.533Z [ERROR] provider.terraform-provider-databricks_v1.10.1: Response contains error diagnostic: diagnostic_detail= diagnostic_summary="inner token: token error: {"error":"invalid_request","error_description":"Identity not found"}" tf_provider_addr=registry.terraform.io/databricks/databricks tf_rpc=PlanResourceChange @module=sdk.proto diagnostic_severity=ERROR tf_proto_version=5.3 tf_req_id=fd335b0e-9fe5-ab1d-d4b5-9e84314e3b72 tf_resource_type=databricks_permissions @caller=/home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/diag/diagnostics.go:55 timestamp=2023-03-01T04:53:20.532Z
2023-03-01T04:53:20.533Z [ERROR] vertex "module.cluster_compute_policy[\"high_memory_cluster\"].databricks_permissions.can_use_cluster_policyinstance_profile" error: inner token: token error: {"error":"invalid_request","error_description":"Identity not found"}
2023-03-01T04:53:20.534Z [ERROR] provider.terraform-provider-databricks_v1.10.1: Response contains error diagnostic: tf_provider_addr=registry.terraform.io/databricks/databricks @caller=/home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/diag/diagnostics.go:55 @module=sdk.proto tf_proto_version=5.3 tf_resource_type=databricks_permissions tf_rpc=PlanResourceChange diagnostic_detail= diagnostic_severity=ERROR diagnostic_summary="inner token: token error: {"error":"invalid_request","error_description":"Identity not found"}" tf_req_id=daf7aaff-9742-782e-30c4-8321d04ecb08 timestamp=2023-03-01T04:53:20.534Z
2023-03-01T04:53:20.534Z [ERROR] vertex "module.cluster_compute_policy[\"general_purpose_cluster\"].databricks_permissions.can_use_cluster_policyinstance_profile" error: inner token: token error: {"error":"invalid_request","error_description":"Identity not found"}
2023-03-01T04:53:20.536Z [ERROR] provider.terraform-provider-databricks_v1.10.1: Response contains error diagnostic: diagnostic_severity=ERROR tf_proto_version=5.3 tf_provider_addr=registry.terraform.io/databricks/databricks tf_resource_type=databricks_permissions @caller=/home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/diag/diagnostics.go:55 @module=sdk.proto diagnostic_detail= diagnostic_summary="inner token: token error: {"error":"invalid_request","error_description":"Identity not found"}" tf_req_id=ba1bdcf6-4ca3-4db9-28e7-91038dd60c37 tf_rpc=PlanResourceChange timestamp=2023-03-01T04:53:20.536Z
2023-03-01T04:53:20.536Z [ERROR] vertex "module.cluster_compute_policy[\"high_compute_cluster\"].databricks_permissions.can_use_cluster_policyinstance_profile" error: inner token: token error: {"error":"invalid_request","error_description":"Identity not found"}
2023-03-01T04:53:20.536Z [ERROR] vertex "module.cluster_compute_policy.databricks_permissions.can_use_cluster_policyinstance_profile (expand)" error: inner token: token error: {"error":"invalid_request","error_description":"Identity not found"}
2023-03-01T04:53:20.537Z [ERROR] vertex "module.cluster_compute_policy.databricks_permissions.can_use_cluster_policyinstance_profile (expand)" error: inner token: token error: {"error":"invalid_request","error_description":"Identity not found"}
2023-03-01T04:53:20.537Z [ERROR] vertex "module.cluster_compute_policy.databricks_permissions.can_use_cluster_policyinstance_profile (expand)" error: inner token: token error: {"error":"invalid_request","error_description":"Identity not found"}
2023-03-01T04:53:20.538Z [INFO]  backend/local: plan operation completed
```

--- a/catalog/resource_grants.go
+++ b/catalog/resource_grants.go
@@ -281,7 +281,7 @@ func ResourceGrants() *schema.Resource {
 		})
 	return common.Resource{
 		Schema: s,
-		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff, c any) error {
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
 			if d.Id() == "" {
 				// unfortunately we cannot do validation before dependent resources exist with tfsdkv2
 				return nil

--- a/catalog/resource_table.go
+++ b/catalog/resource_table.go
@@ -85,7 +85,7 @@ func ResourceTable() *schema.Resource {
 		"view_definition", "comment", "properties"})
 	return common.Resource{
 		Schema: tableSchema,
-		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff, c any) error {
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
 			if d.Get("table_type") != "EXTERNAL" {
 				return nil
 			}

--- a/common/resource.go
+++ b/common/resource.go
@@ -20,7 +20,7 @@ type Resource struct {
 	Read           func(ctx context.Context, d *schema.ResourceData, c *DatabricksClient) error
 	Update         func(ctx context.Context, d *schema.ResourceData, c *DatabricksClient) error
 	Delete         func(ctx context.Context, d *schema.ResourceData, c *DatabricksClient) error
-	CustomizeDiff  func(ctx context.Context, d *schema.ResourceDiff, c any) error
+	CustomizeDiff  func(ctx context.Context, d *schema.ResourceDiff) error
 	StateUpgraders []schema.StateUpgrader
 	Schema         map[string]*schema.Schema
 	SchemaVersion  int
@@ -50,6 +50,32 @@ func recoverable(cb func(
 			}
 		}()
 		err = cb(ctx, d, c)
+		return
+	}
+}
+
+func (r Resource) saferCustomizeDiff() schema.CustomizeDiffFunc {
+	if r.CustomizeDiff == nil {
+		return nil
+	}
+	return func(ctx context.Context, rd *schema.ResourceDiff, _ any) (err error) {
+		defer func() {
+			// this is deliberate decision to convert a panic into error,
+			// so that any unforeseen bug would we visible to end-user
+			// as an error and not a provider crash, which is way less
+			// of pleasant experience.
+			if panic := recover(); panic != nil {
+				err = nicerError(ctx, fmt.Errorf("panic: %v", panic),
+					"customize diff for")
+			}
+		}()
+		// we don't propagate instance of SDK client to the diff function, because
+		// authentication is not deterministic at this stage with the recent Terraform
+		// versions. Diff customization must be limited to hermetic checks only anyway.
+		err = r.CustomizeDiff(ctx, rd)
+		if err != nil {
+			err = nicerError(ctx, err, "customize diff for")
+		}
 		return
 	}
 }
@@ -117,7 +143,7 @@ func (r Resource) ToResource() *schema.Resource {
 		Schema:         r.Schema,
 		SchemaVersion:  r.SchemaVersion,
 		StateUpgraders: r.StateUpgraders,
-		CustomizeDiff:  r.CustomizeDiff,
+		CustomizeDiff:  r.saferCustomizeDiff(),
 		CreateContext: func(ctx context.Context, d *schema.ResourceData,
 			m any) diag.Diagnostics {
 			c := m.(*DatabricksClient)

--- a/common/resource_test.go
+++ b/common/resource_test.go
@@ -155,3 +155,26 @@ func TestRecoverableFromPanic(t *testing.T) {
 	assert.True(t, diags.HasError())
 	assert.Equal(t, "cannot read sample: panic: what to do?...", diags[0].Summary)
 }
+
+func TestCustomizeDiffRobustness(t *testing.T) {
+	r := Resource{
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
+			return fmt.Errorf("nope")
+		},
+	}.ToResource()
+
+	ctx := context.Background()
+	ctx = context.WithValue(ctx, ResourceName, "sample")
+
+	err := r.CustomizeDiff(ctx, nil, nil)
+	assert.EqualError(t, err, "cannot customize diff for sample: nope")
+
+	r = Resource{
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
+			panic("oops")
+		},
+	}.ToResource()
+
+	err = r.CustomizeDiff(ctx, nil, nil)
+	assert.EqualError(t, err, "cannot customize diff for sample: panic: oops")
+}

--- a/jobs/resource_job.go
+++ b/jobs/resource_job.go
@@ -595,7 +595,7 @@ func ResourceJob() *schema.Resource {
 			Create: schema.DefaultTimeout(clusters.DefaultProvisionTimeout),
 			Update: schema.DefaultTimeout(clusters.DefaultProvisionTimeout),
 		},
-		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff, m any) error {
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
 			var js JobSettings
 			common.DiffToStructPointer(d, jobSchema, &js)
 			alwaysRunning := d.Get("always_running").(bool)

--- a/mws/resource_mws_workspaces.go
+++ b/mws/resource_mws_workspaces.go
@@ -570,7 +570,7 @@ func ResourceMwsWorkspaces() *schema.Resource {
 			}
 			return NewWorkspacesAPI(ctx, c).Delete(accountID, workspaceID)
 		},
-		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff, m any) error {
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
 			old, new := d.GetChange("private_access_settings_id")
 			if old != "" && new == "" {
 				return fmt.Errorf("cannot remove private access setting from workspace")

--- a/permissions/resource_permissions.go
+++ b/permissions/resource_permissions.go
@@ -4,7 +4,6 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"log"
 	"path"
 	"strconv"
 	"strings"
@@ -381,21 +380,7 @@ func ResourcePermissions() *schema.Resource {
 	})
 	return common.Resource{
 		Schema: s,
-		CustomizeDiff: func(ctx context.Context, diff *schema.ResourceDiff, c any) error {
-			client := c.(*common.DatabricksClient)
-			log.Printf("[DEBUG] permissions id=%s, config_present=%v", diff.Id(), client.Config != nil)
-			if client.Config.Host == "" || client.DatabricksClient.Config.Host == "" {
-				log.Printf("[WARN] cannot validate permission levels, because host is not known yet")
-				return nil
-			}
-			w, err := client.WorkspaceClient()
-			if err != nil {
-				return err
-			}
-			me, err := w.CurrentUser.Me(ctx)
-			if err != nil {
-				return fmt.Errorf("customize diff: me: %w", err)
-			}
+		CustomizeDiff: func(ctx context.Context, diff *schema.ResourceDiff) error {
 			// Plan time validation for object permission levels
 			for _, mapping := range permissionsResourceIDFields() {
 				if _, ok := diff.GetOk(mapping.field); !ok {
@@ -406,10 +391,8 @@ func ResourcePermissions() *schema.Resource {
 					m := access_control.(map[string]any)
 					permission_level := m["permission_level"].(string)
 					if !stringInSlice(permission_level, mapping.allowedPermissionLevels) {
-						return fmt.Errorf(`permission_level %s is not supported with %s objects`, permission_level, mapping.field)
-					}
-					if m["user_name"].(string) == me.UserName {
-						return fmt.Errorf("it is not possible to decrease administrative permissions for the current user: %s", me.UserName)
+						return fmt.Errorf(`permission_level %s is not supported with %s objects`,
+							permission_level, mapping.field)
 					}
 				}
 			}
@@ -443,12 +426,25 @@ func ResourcePermissions() *schema.Resource {
 		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
 			var entity PermissionsEntity
 			common.DataToStructPointer(d, s, &entity)
+			w, err := c.WorkspaceClient()
+			if err != nil {
+				return err
+			}
+			me, err := w.CurrentUser.Me(ctx)
+			if err != nil {
+				return err
+			}
+			// this logic was moved from CustomizeDiff because of undeterministic auth behavior
+			// in the corner-case scenarios.
+			// see https://github.com/databricks/terraform-provider-databricks/issues/2052
+			for _, v := range entity.AccessControlList {
+				if v.UserName == me.UserName {
+					format := "it is not possible to decrease administrative permissions for the current user: %s"
+					return fmt.Errorf(format, me.UserName)
+				}
+			}
 			for _, mapping := range permissionsResourceIDFields() {
 				if v, ok := d.GetOk(mapping.field); ok {
-					w, err := c.WorkspaceClient()
-					if err != nil {
-						return err
-					}
 					id, err := mapping.idRetriever(ctx, w, v.(string))
 					if err != nil {
 						return err

--- a/permissions/resource_permissions_test.go
+++ b/permissions/resource_permissions_test.go
@@ -409,51 +409,32 @@ func TestResourcePermissionsRead_some_error(t *testing.T) {
 	assert.Error(t, err)
 }
 
-func TestResourcePermissionsCustomizeDiff_ErrorOnScimMe(t *testing.T) {
+func TestResourcePermissionsCustomizeDiff_ErrorOnCreate(t *testing.T) {
+	qa.ResourceFixture{
+		Resource: ResourcePermissions(),
+		Create:   true,
+		HCL: `
+		cluster_id = "abc"
+		access_control {
+			permission_level = "WHATEVER"
+		}`,
+	}.ExpectError(t, "permission_level WHATEVER is not supported with cluster_id objects")
+}
+
+func TestResourcePermissionsCustomizeDiff_ErrorOnPermissionsDecreate(t *testing.T) {
 	qa.ResourceFixture{
 		Fixtures: []qa.HTTPFixture{
-			{
-				Method:   http.MethodGet,
-				Resource: "/api/2.0/permissions/clusters/abc",
-				Response: ObjectACL{
-					ObjectID:   "/clusters/abc",
-					ObjectType: "clusters",
-					AccessControlList: []AccessControl{
-						{
-							UserName: TestingUser,
-							AllPermissions: []Permission{
-								{
-									PermissionLevel: "CAN_READ",
-									Inherited:       false,
-								},
-							},
-						},
-						{
-							UserName: TestingAdminUser,
-							AllPermissions: []Permission{
-								{
-									PermissionLevel: "CAN_MANAGE",
-									Inherited:       false,
-								},
-							},
-						},
-					},
-				},
-			},
-			{
-				Method:   http.MethodGet,
-				Resource: "/api/2.0/preview/scim/v2/Me",
-				Response: apierr.APIErrorBody{
-					ErrorCode: "INVALID_REQUEST",
-					Message:   "Internal error happened",
-				},
-				Status: 400,
-			},
+			me,
 		},
 		Resource: ResourcePermissions(),
-		Read:     true,
-		ID:       "/clusters/abc",
-	}.ExpectError(t, "customize diff: me: Internal error happened")
+		Create:   true,
+		HCL: `
+		cluster_id = "abc"
+		access_control {
+			permission_level = "CAN_ATTACH_TO"
+			user_name = "admin"
+		}`,
+	}.ExpectError(t, "it is not possible to decrease administrative permissions for the current user: admin")
 }
 
 func TestResourcePermissionsRead_ErrorOnScimMe(t *testing.T) {

